drivers/note/noteram_driver.c

```diff
@@ -435,6 +586,17 @@ void sched_note_add(FAR const void *note, size_t notelen)
       next = noteram_next(head, 1);
       if (next == g_noteram_info.ni_tail)
         {
+          if (!g_noteram_info.ni_overwrite)
+            {
+              /* Stop recording if not in overwrite mode */
+
+#ifdef CONFIG_SMP
+              spin_unlock_wo_note(&g_noteram_lock);
+              up_irq_restore(flags);
+#endif
+              return;
```

Comment:
should we break here to otherwise we will forget to update ni_head at line 451?

Reply:
In this case ni_head should not be updated. Because an incomplete note struct is recorded in the circular buffer if updated. If not updated, the incompletely recorded note will be simply discarded. 

But, I found another problem in this case.
If the buffer have 16bytes free space, and 20bytes note and 12bytes note arrives sequentially, the former incoming note is discarded but only the latter note is recorded. This is an unexpected behavior.
I think that we need to add the flag to control whether we record the incoming note or not.
When in not overwrite mode, once we found an incoming note which is greater than the free space, disable the flag to block the following notes.

Reply:
> In this case ni_head should not be updated. Because an incomplete note struct is recorded in the circular buffer if updated. If not updated, the incompletely recorded note will be simply discarded.
> 

Ok.

> But, I found another problem in this case.
> If the buffer have 16bytes free space, and 20bytes note and 12bytes note arrives sequentially, the former incoming note is discarded but only the latter note is recorded. This is an unexpected behavior.
> I think that we need to add the flag to control whether we record the incoming note or not.
> When in not overwrite mode, once we found an incoming note which is greater than the free space, disable the flag to block the following notes.

maybe we can change ni_overwrite to tristate(disable, enable, overflow).